core/src/SYCL/Kokkos_SYCL_UniqueToken.hpp

```diff
@@ -81,51 +86,80 @@ class UniqueToken<SYCL, UniqueTokenScope::Global> {
 
   /// \brief upper bound for acquired values, i.e. 0 <= value < size()
   KOKKOS_INLINE_FUNCTION
-  size_type size() const noexcept { return m_count; }
+  size_type size() const noexcept { return m_locks.extent(0); }
 
+ protected:
+  // Constructors for the Instance version
+  UniqueToken(size_type max_size,
+              execution_space const& arg = execution_space())
+      : m_locks(Kokkos::View<uint32_t*, SYCLDeviceUSMSpace>(
+            Kokkos::view_alloc(arg, "Kokkos::UniqueToken::m_locks"),
+            max_size)) {}
+
+ private:
   /// \brief acquire value such that 0 <= value < size()
   KOKKOS_INLINE_FUNCTION
-  size_type acquire() const {
-    const Kokkos::pair<int, int> result =
-        Kokkos::Impl::concurrent_bitset::acquire_bounded(
-            m_buffer, m_count
-#ifdef KOKKOS_ARCH_INTEL_GPU
-            ,
-            Kokkos::Impl::clock_tic() % m_count
-#endif
-        );
-
-    if (result.first < 0) {
-      Kokkos::abort(
-          "UniqueToken<SYCL> failure to acquire tokens, no tokens available");
+  size_type impl_acquire() const {
+    auto item = sycl::ext::oneapi::experimental::this_nd_item<3>();
+    std::size_t threadIdx[3] = {item.get_local_id(2), item.get_local_id(1),
+                                item.get_local_id(0)};
+    std::size_t blockIdx[3]  = {item.get_group(2), item.get_group(1),
+                               item.get_group(0)};
+    std::size_t blockDim[3] = {item.get_local_range(2), item.get_local_range(1),
+                               item.get_local_range(0)};
+
+    int idx = (blockIdx[0] * (blockDim[0] * blockDim[1]) +
+               threadIdx[1] * blockDim[0] + threadIdx[0]) %
+              size();
```

Comment:
I would prefer if you split the expression in two and had `idx = idx % size();` like we do in CUDA